### PR TITLE
CAF-3648: Job Service Docs

### PR DESCRIPTION
**Pull request by glucy**
https://jira.autonomy.com/browse/CAF-3648

Document changes to the Job Service documentation in relation to supporting "dependent jobs".

### CODE REVIEW

**Comment by glucy** [comment]
SEPG Build of the Docs:  https://pages.github.houston.softwaregrp.net/sepg-docs-qa/JobService_CI_job-service_CAF-3648_Job_Service_Docs/pages/en-us/Architecture

File: docs/pages/en-us/Architecture.md [hunk 1]
```diff
@@ -100,6 +99,8 @@ When the job tracking worker receives a success message to be proxied (that is,
 
 The job tracking worker recognizes failure and retry messages, which are being proxied, and updates the job database accordingly.
 
+The job tracking worker can also automatically forward on dependent jobs for execution.  A dependent job is a job which must wait until a specific job or list of jobs have completed before it, itself can be executed.  As the job tracking worker monitors a job's progress, should a job complete, the job tracking worker will receive a list of jobs which can now be executed as the job(s) which these jobs were dependent upon have now Completed.
```
Comment (michael-bryson) [inline docs/pages/en-us/Architecture.md]: > specific job or list of jobs have completed before it, itself can be executed.

I don't think you need the word itself or the comma here:

> specific job or list of jobs have completed before it can be executed.

Comment (michael-bryson) [reply docs/pages/en-us/Architecture.md]: > As the job tracking worker monitors a job's progress, should a job complete, the job tracking worker will receive a list of jobs which can now be executed as the job(s) which these jobs were dependent upon have now Completed.

I would rewrite this as:

> The job tracking worker monitors a job's progress, when a job completes, the job tracking worker will receive a list of jobs which can now be executed as the job(s) which these jobs were dependent upon have now completed.

File: docs/pages/en-us/Architecture.md [hunk 2]
```diff
@@ -125,3 +126,26 @@ This table stores information on the jobs that are requested. Entries are added
 The task tables have the same structure as the job table. Each job has one task table, which is created when the first subtask is reported, and deleted when the job completes successfully. If the job fails, the table is retained for a period of time for examination.
 
 When a task is marked complete, the system checks whether the parent task (or the job if it is the top level) can also be marked complete.
+
+### Job Dependency Table
+
+This table stories Ids of dependent jobs.  Jobs which must be completed before the job in question can be executed.
```
Comment (michael-bryson) [inline docs/pages/en-us/Architecture.md]: "stores" not "stories"

Comment (michael-bryson) [reply docs/pages/en-us/Architecture.md]: Maybe reword this to:

> This table stores Ids of dependent jobs i.e. jobs which must be completed before the job in question can be executed.

File: docs/pages/en-us/Architecture.md [hunk 3]
```diff
@@ -125,3 +126,26 @@ This table stores information on the jobs that are requested. Entries are added
...
+
+| **Column**        | **Data Type** | **Nullable?** | **Primary Key?** |
+|-------------------|---------------|---------------|------------------|
+| Job_Id            | String        | No            | Yes              |
+| Dependent_Job_Id  | String        | No            |                  |
+
+### Job Task Data
+
+This table stores information on jobs which have dependent jobs and must waiting for execution.  The table contains enough information for the job tracking worker to forward on the job once it's dependent jobs have all completed.
```
Comment (michael-bryson) [inline docs/pages/en-us/Architecture.md]: Reword this one:

> This table stores information on jobs which have dependent jobs and must **wait** for execution.  The table contains enough information for the job tracking worker to forward on the job, once it's dependent jobs have all completed.

Note the added comma.

File: worker-jobtracking/README.md [hunk 4]
```diff
@@ -7,3 +7,5 @@ Messages will typically arrive at the Job Tracking Worker because the queue that
 trackingPipe in a task message, which will trigger the Worker Framework to re-route the message to the Job Tracking Worker. The worker
 checks for task cancellation, reports the progress of the task to the Job Database, and forwards the message to the correct destination
 queue, e.g. the input queue of another worker.
+
+The Job Service has the ability to execute jobs in an order defined by a job having a dependency on another job.  The Job Service will not execute a job until all the jobs it is dependent upon have completed.  As the job tracking worker monitors the progress of jobs, once all dependent jobs have been completed the job tracking worker will automatically forward on the jobs for execution which were previous waiting upon dependent jobs.
```
Comment (michael-bryson) [inline worker-jobtracking/README.md]: Remove the "As" at the start of the second sentence.

"previously" not "previous"

**Review by michael-bryson — CHANGES_REQUESTED** [review]

**Comment by glucy** [comment]
I agree with Michael's comments & have made grammar updates.

File: docs/pages/en-us/Architecture.md [hunk 5]
```diff
@@ -125,3 +126,26 @@ This table stores information on the jobs that are requested. Entries are added
...
+
+| **Column**        | **Data Type** | **Nullable?** | **Primary Key?** |
+|-------------------|---------------|---------------|------------------|
+| Job_Id            | String        | No            | Yes              |
+| Task_Classifier   | String        | No            |                  |
+| Task_Api_Version  | String        | No            |                  |
+| Task_Data         | String        | No            |                  |
```
Comment (CJComac) [inline docs/pages/en-us/Architecture.md]: Task_Api_Version is an integer and Task_Data is a bytea (aka byte array in postgres)

**Review by michael-bryson — APPROVED** [review]

**Comment by glucy** [comment]
@dermot-hardy Do we need to update the Architecture Diagram for the JobService **or** should we create a new separate Architecture Diagram to cover Dependent Jobs **or** leave things as they are for now?

Please bare in mind we are unable to edit the existing diagram and it will mean starting again.  (Conal created the original diagram and I've been unable to find a tool that will allow it to easily editable).

File: docs/pages/en-us/Overview.md [hunk 6]
```diff
@@ -36,6 +36,9 @@ The batch worker takes batches of work, recursively breaks them down into ever s
 
 The Job Service is particularly beneficial to the batch worker because it monitors the progress of the batch as a whole. Even if the size of the batch is not known in advance, using the Job Service makes it possible to monitor the overall progress of the entire batch, as well as making other operations available, such as the capability to cancel the batch.
 
+### Dependent Jobs
+The Job Service has the ability to execute jobs in an order defined by a job having a dependency on another job.  The Job Service will not execute a job until all the jobs it is dependent upon have completed.  Once all dependent jobs have completed the Job Service will automatically execute the job which was previously waiting upon dependent jobs.
```
Comment (CJComac) [inline docs/pages/en-us/Overview.md]: The Job Tracking Worker is the component that automatically starts a job when its prerequisites have been completed.

File: docs/pages/en-us/Overview.md [hunk 7]
```diff
@@ -36,6 +36,9 @@ The batch worker takes batches of work, recursively breaks them down into ever s
 
 The Job Service is particularly beneficial to the batch worker because it monitors the progress of the batch as a whole. Even if the size of the batch is not known in advance, using the Job Service makes it possible to monitor the overall progress of the entire batch, as well as making other operations available, such as the capability to cancel the batch.
 
+### Dependent Jobs
```
Comment (CJComac) [inline docs/pages/en-us/Overview.md]: Move Dependent Jobs section below the Example Usages section.

**Review by CJComac — CHANGES_REQUESTED** [review]
@glucy 
- I think there should be a 12th bullet point on the architecture diagram illustrating that the Job Tracking Worker will start all jobs that are dependent on the job that just completed and who's prerequisite jobs have been completed.
- The Getting Started guide needs adjusted to at least illustrate how to define a job with prerequisite jobs?

File: docs/pages/en-us/Overview.md (hunk 7 above)
Comment (glucy) [reply docs/pages/en-us/Overview.md]: Surely examples always come last?  Describe everything first then the examples?

File: docs/pages/en-us/Overview.md (hunk 6 above)
Comment (dermot-hardy) [reply docs/pages/en-us/Overview.md]: @CJComac That is an irrelevant implementation detail (from a user perspective).

Comment (dermot-hardy) [reply docs/pages/en-us/Overview.md]: "The Job Service has the ability to execute jobs in an order defined by a job having a dependency on another job.": This doesn't seem very clear to me.

Comment (glucy) [reply docs/pages/en-us/Overview.md]: I can update to the below or leave as is.  We are describing the "Job Service" as a whole unit here I guess.

`Once all dependent jobs have completed the job tracking worker will automatically forward the job which was previously waiting upon dependent jobs.`

File: docs/pages/en-us/Architecture.md (hunk 5 above)
Comment (glucy) [reply docs/pages/en-us/Architecture.md]: Updated

File: docs/pages/en-us/Architecture.md [hunk 8]
```diff
@@ -100,6 +99,8 @@ When the job tracking worker receives a success message to be proxied (that is,
 
 The job tracking worker recognizes failure and retry messages, which are being proxied, and updates the job database accordingly.
 
+The job tracking worker can also automatically forward on dependent jobs for execution.  A dependent job is a job which must wait until a specific job or list of jobs have completed before it can be executed.  The job tracking worker monitors a job's progress, when a job completes, the job tracking worker will receive a list of jobs which can now be executed as the job(s) which these jobs were dependent upon have now completed.
```
Comment (dermot-hardy) [inline docs/pages/en-us/Architecture.md]: The last sentence in this paragraph doesn't read well.

File: docs/pages/en-us/Architecture.md [hunk 9]
```diff
@@ -125,3 +126,26 @@ This table stores information on the jobs that are requested. Entries are added
...
 
 When a task is marked complete, the system checks whether the parent task (or the job if it is the top level) can also be marked complete.
+
+### Job Dependency Table
+
+This table stores Ids of dependent jobs i.e. jobs which must be completed before the job in question can be executed.
+
+| **Column**        | **Data Type** | **Nullable?** | **Primary Key?** |
+|-------------------|---------------|---------------|------------------|
+| Job_Id            | String        | No            | Yes              |
```
Comment (dermot-hardy) [inline docs/pages/en-us/Architecture.md:137]: It doesn't make sense for Job_Id to be the primary key of this table.

**Review by dermot-hardy — APPROVED** [review]

**Comment by dermot-hardy** [comment]
@glucy I think the diagram is fine as it is.  Perhaps add some extra sub-points to point 9 to explain the extra stuff going on.

File: docs/pages/en-us/Architecture.md (hunk 8 above)
Comment (glucy) [reply docs/pages/en-us/Architecture.md]: Will update to:
`The job tracking worker monitors a job's progress, when a job completes, the job tracking worker will receive a list of jobs which can now be executed.`

File: docs/pages/en-us/Architecture.md (hunk 9 above)
Comment (glucy) [reply docs/pages/en-us/Architecture.md:137]: Primary Key should be `Yes` for both columns

File: docs/pages/en-us/Overview.md (hunk 6 above)
Comment (glucy) [reply docs/pages/en-us/Overview.md]: Leaving as it is i.e. `Once all dependent jobs have completed the Job Service will automatically execute the job which was previously waiting upon dependent jobs.`

**Comment by glucy** [comment]
@dermot-hardy Added an additional sub point to the point 9 of the diagram annotation:
`If a job completes and has associated dependent jobs, the job tracking worker receives a list of jobs now available for execution. The job tracking worker forwards these jobs to the target pipe.`

**Review by CJComac — APPROVED** [review]
